### PR TITLE
Feature/tokenomics todos

### PR DESCRIPTION
**Pull request by MilGard91**
## Context

Include relevant motivation and context. A link to the issue is enough if it contains all the relevant 
information.

## Changes proposed in this pull request

Include a summary of the change.

## Test plan

Please describe the tests cases that you ran to verify your changes. Add further instructions on 
how to run them if needed (i.e. migration / deployment scripts, env vars, etc).

## Checklist

- [ ] Rebased the base branch
- [ ] Attached corresponding Github issue
- [ ] Prefixed the name with the type of change (i.e. feat, chore, test)
- [ ] Performed a self-review of my own code
- [ ] Followed the style guidelines of this project
- [ ] Made corresponding changes to the documentation
- [ ] Didn't generate new warnings
- [ ] Didn't generate failures on existing tests
- [ ] Added tests that prove my fix is effective or that my feature works

## Review

When reviewing a PR, please indicate intention in comments using the following emojis:
* :cake: = Nice to have but not essential.
* :bulb: = Suggestion or a comment based on personal opinion
* :hammer: = I believe this should be changed.
* :thinking: = I don’t understand something, do you mind giving me more context?
* :rocket: = Feedback

### CODE REVIEW

File: test/unit/Ramm/swap.js [hunk 1]
```diff
@@ -24,7 +24,21 @@ describe('swap', function () {
...
+  it('should revert if block timestamp surpasses deadline', async function () {
+    const fixture = await loadFixture(setup);
+    const { ramm } = fixture.contracts;
+    const [member] = fixture.accounts.members;
+
+    const nxmIn = parseEther('1');
+    const minTokensOut = parseEther('0.015'); // 0.0152 ETH initial spot price
+    const { timestamp } = await ethers.provider.getBlock('latest');
+    const deadline = timestamp + 4 * 60; // add 5 minutes
+    setNextBlockTime(timestamp + 5 * 60);
```
Comment (shark0der) [inline test/unit/Ramm/swap.js:39]: No need to advance time - just subtract 1 second from the current timestamp.

File: test/unit/Ramm/swap.js [hunk 2]
```diff
@@ -35,7 +49,11 @@ describe('swap', function () {
     const ethIn = parseEther('1');
     const minTokensOut = parseEther('29'); // 1ETH = 28.8NXM at 0.0347ETH
 
-    await expect(ramm.connect(member).swap(0, minTokensOut, { value: ethIn })).to.be.revertedWith(
+    const { timestamp } = await ethers.provider.getBlock('latest');
+    const deadline = timestamp + 5 * 60; // add 5 minutes
+    setNextBlockTime(timestamp + 4 * 60);
```
Comment (shark0der) [inline test/unit/Ramm/swap.js:54]: No need to advance time at all.

File: contracts/modules/capital/Ramm.sol [hunk 3]
```diff
@@ -145,15 +146,23 @@ contract Ramm is IRamm, MasterAwareV2 {
...
+    // update TWAP
+    {
+      Observation[3] memory _observations = observations;
+      _observations = _updateTwap(initialState, _observations, block.timestamp, capital, supply, mcrValue);
+      for (uint i = 0; i < _observations.length; i++) {
+        observations[i] = _observations[i];
+      }
+    }
+
+  uint nxmA = state.nxmA;
```
Comment (shark0der) [inline contracts/modules/capital/Ramm.sol:165]: indent :eyes:

Comment (shark0der) [inline contracts/modules/capital/Ramm.sol]: Why did you move this up?

Comment (MilGard91) [reply contracts/modules/capital/Ramm.sol]: Stack too deep

File: contracts/modules/capital/Ramm.sol [hunk 4]
```diff
@@ -214,7 +221,12 @@ contract Ramm is IRamm, MasterAwareV2 {
     if (eth < TARGET_LIQUIDITY) {
       // inject eth
       uint timeLeftOnBudget = budget * LIQ_SPEED_PERIOD / FAST_LIQUIDITY_SPEED;
-      uint maxToInject = TARGET_LIQUIDITY - eth;
+      uint maxToInject;
+      if (capital <= mcrValue + TARGET_LIQUIDITY) {
+        maxToInject = 0;
+      } else {
+        maxToInject = Math.min(TARGET_LIQUIDITY - eth, capital - mcrValue - TARGET_LIQUIDITY);
```
Comment (shark0der) [inline contracts/modules/capital/Ramm.sol:228]: `capital - mcrValue - TARGET_LIQUIDITY` may underflow

File: contracts/modules/capital/Ramm.sol [hunk 5]
```diff
@@ -258,8 +270,8 @@ contract Ramm is IRamm, MasterAwareV2 {
         nxmA = eth * supply / bufferedCapitalA;
       } else {
         // use ratchet
-        uint nr_denom_addend = r * capital * nxmA / supply / RATCHET_PERIOD / RATCHET_DENOMINATOR;
-        nxmA = eth * nxmA / (eth - nr_denom_addend);
+        //   = eth * nxmA / (eth - nr_denom_addend)
```
Comment (shark0der) [inline contracts/modules/capital/Ramm.sol:273]: this comment can be removed now as it's more confusing than helpful

File: contracts/modules/capital/Ramm.sol [hunk 6]
```diff
@@ -278,8 +290,8 @@ contract Ramm is IRamm, MasterAwareV2 {
         nxmB = eth * supply / bufferedCapitalB;
       } else {
         // use ratchet
-        uint nr_denom_addend = nxmB * elapsed * state.ratchetSpeed * capital / supply / RATCHET_PERIOD / RATCHET_DENOMINATOR;
-        nxmB = eth * nxmB / (eth + nr_denom_addend);
+        //   = eth * nxmB / (eth + nr_denom_addend)
```
Comment (shark0der) [inline contracts/modules/capital/Ramm.sol:293]: Same

File: contracts/modules/capital/Ramm.sol [hunk 7]
```diff
@@ -165,7 +174,6 @@ contract Ramm is IRamm, MasterAwareV2 {
     uint ethOut = state.eth - eth;
     require(ethOut >= minTokensOut, "Ramm: ethOut is less than minTokensOut");
 
-    // TODO add buffer into calculation
     require(capital - ethOut >= mcr().getMCR(), "NO_SWAPS_IN_BUFFER_ZONE");
```
Comment (shark0der) [inline contracts/modules/capital/Ramm.sol:177]: Redundant getMCR call, we already have the value from parameters

Comment (shark0der) [inline contracts/modules/capital/Ramm.sol:177]: Redundant getMCR call, we already have the value from parameters

**Review by shark0der — APPROVED** [review]
